### PR TITLE
Add functionality to download files in parallel

### PR DESCRIPTION
**Pull request by jsm**
@alex-optimizely @onufryk @4vanger

### CODE REVIEW

File: pistachio/s3.py [hunk 1]
```diff
@@ -24,16 +28,52 @@ def download(conn, bucket, path=[]):
...
+          thread = threading.Thread(target=fetch_config_partial, args=(folder, key))
+          thread.start()
+          threads.append(thread)
+        else:
+          fetch_config_partial(folder, key)
+
+  # Wait for the threads to finish if we're running in parallel
+  if parallel:
+    for thread in threads:
+      thread.join()
```
Comment (jordangarcia) [inline pistachio/s3.py]: this imperatively blocks until all threads finish?

Comment (jsm) [inline pistachio/s3.py]: righto

File: pistachio/s3.py [hunk 2]
```diff
@@ -24,16 +28,52 @@ def download(conn, bucket, path=[]):
...
+  # Reset the config_partials array
+  global config_partials
+  config_partials = {}
+  # Must store partials by folder, so that we guarantee folder hierarchy when merging them
+  for folder in path:
+    config_partials[folder] = []
+
+  # Create thread store if we're running in parallel
+  if parallel:
+    threads = []
```
Comment (jordangarcia) [inline pistachio/s3.py]: is this safe in python, ie is this variable hoisted to the proper scope such that it can be accessed later?

**Comment by jordangarcia** [comment]
few questions regarding python scope / concurrency lgtm

File: pistachio/s3.py (hunk 2 above)
Comment (jsm) [inline pistachio/s3.py]: Right. Python if blocks have the same scope as the scope they are called in.

**Comment by jsm** [comment]
FYI Planning on adding some integration tests in the future (using a live bucket)

File: pistachio/s3.py [hunk 3]
```diff
@@ -24,16 +28,52 @@ def download(conn, bucket, path=[]):
...
-          # Update the config with the config partial
-          config_partial = yaml.load(contents)
-          util.merge_dicts(config, config_partial)
-        except boto.exception.S3ResponseError:
-          pass # Access denied. Skip this one.
+        # Download and store
+        if parallel:
+          thread = threading.Thread(target=fetch_config_partial, args=(folder, key))
+          thread.start()
+          threads.append(thread)
```
Comment (bhamodi) [inline pistachio/s3.py]: Should we have a limit on max threads?

Comment (jsm) [inline pistachio/s3.py]: Nah, they're network intensive, not cpu intensive

**Comment by jsm** [comment]
Comitted by https://github.com/optimizely/pistachio/commit/175017d8d1183ba5e56330d29bf4d7ef5e1296f4

**Pull request closed without merging**